fdbserver/include/fdbserver/TenantEntryCache.actor.h

```diff
@@ -342,24 +342,53 @@ class TenantEntryCache : public ReferenceCounted<TenantEntryCache<T>>, NonCopyab
 
 	void put(const TenantNameEntryPair& pair) {
 		TenantEntryCachePayload<T> payload = createPayloadFunc(pair.first, pair.second);
-		if (mapByTenantId.find(pair.second.id) == mapByTenantId.end()) {
-			ASSERT(mapByTenantName.find(pair.first) == mapByTenantName.end());
+		auto idItr = mapByTenantId.find(pair.second.id);
+		auto nameItr = mapByTenantName.find(pair.first);
 
+		if (idItr == mapByTenantId.end() && nameItr == mapByTenantName.end()) {
+			// New tenant entry
 			mapByTenantId[pair.second.id] = payload;
 			mapByTenantName[pair.first] = payload;
 
 			TraceEvent(SevInfo, "TenantEntryCacheNewTenant", uid)
 			    .detail("TenantName", pair.first.contents().toString())
 			    .detail("TenantID", pair.second.id)
 			    .detail("TenantPrefix", pair.second.prefix);
-		} else {
+
+			CODE_PROBE(true, "TenantCache new entry");
+		} else if (idItr != mapByTenantId.end() && nameItr == mapByTenantName.end()) {
+			// Tenant exists with a given TenantId, but, with a different name i.e. Tenant renamed
+			TenantName existing = idItr->value.name;
+			mapByTenantName.erase(existing);
+
+			mapByTenantId[pair.second.id] = payload;
+			mapByTenantName[pair.first] = payload;
+
+			TraceEvent(SevInfo, "TenantEntryCacheUpdateTenantName", uid)
+			    .detail("TenantName", pair.first.contents().toString())
+			    .detail("TenantNameExisting", existing.contents().toString())
+			    .detail("TenantID", pair.second.id)
+			    .detail("TenantPrefix", pair.second.prefix);
+
+			CODE_PROBE(true, "TenantCache entry TenantName updated");
+		} else if (idItr == mapByTenantId.end() && nameItr != mapByTenantName.end()) {
+			// Tenant exists with same name, but, different TenantId (delete and recreate the tenant)
+			int64_t existing = nameItr->value.entry.id;
+
+			mapByTenantId.erase(existing);
+
 			mapByTenantId[pair.second.id] = payload;
 			mapByTenantName[pair.first] = payload;
 
-			TraceEvent(SevInfo, "TenantEntryCacheUpdateTenant", uid)
+			TraceEvent(SevInfo, "TenantEntryCacheUpdateTenantId", uid)
 			    .detail("TenantName", pair.first.contents().toString())
+			    .detail("TenantIdExisting", existing)
 			    .detail("TenantID", pair.second.id)
 			    .detail("TenantPrefix", pair.second.prefix);
+
+			CODE_PROBE(true, "TenantCache entry TenantId updated");
+		} else {
+			// Tenant already exists, do nothing
```

Comment:
Two notes in this case -- one is that it's possible the tenant could have changed in other properties (e.g. it could have moved tenant groups), so we may still want to allow updating the cache even if the ID and name are the same as before. 

The second thing is that it's possible that some combination of renaming and creating/deleting new tenants could result in both entries being present but not being associated with each other. In other words, if we specify a tenant here with name `foo` and ID 1, those may not match the same tenant in the cache. `foo` could have been renamed from `bar` (so `bar` has ID 1 in the cache) and there may have been a previous tenant `foo` with ID 0 that was deleted. That means we'll need to verify that these entries are the same and clean them up first if not.

Reply:
> one is that it's possible the tenant could have changed in other properties (e.g. it could have moved tenant groups), so we may still want to allow updating the cache even if the ID and name are the same as before.

I did think about it (actually had it and removed it), the way entry is inserted in the cache is by invoking the registered callback routine. I could add it back if we assume supplying the same `name` and `TenantMapEntry` can still generate different `payload`

Reply:
>The second thing is that it's possible that some combination of renaming and creating/deleting new tenants could result in both entries being present but not being associated with each other. In other words, if we specify a tenant here with name foo and ID 1, those may not match the same tenant in the cache. foo could have been renamed from bar (so bar has ID 1 in the cache) and there may have been a previous tenant foo with ID 0 that was deleted. That means we'll need to verify that these entries are the same and clean them up first if not.

Such cases should be handled when the cache gets refreshed in the next cycle, assuming those cases are rare enough, not sure if cache bloat is something that needs to concerned about.

Reply:
> Such cases should be handled when the cache gets refreshed in the next cycle, assuming those cases are rare enough, not sure if cache bloat is something that needs to concerned about.

The problem here is that we are choosing not to cache the entry in this case. Essentially the user is supplying an entry unknown to the cache, but the cache ignores it because it is misled into thinking that the entry is already present. I could see an argument that we don't need to clean up the stale pairs (i.e. if we match on an ID, it may not be required to delete the old entry from the name map, although we might as well given what we know), but I think it would be surprising if the cache failed to insert the new entry you supplied.

Reply:
> The problem here is that we are choosing not to cache the entry in this case. Essentially the user is supplying an entry unknown to the cache,

Inputs to the `put` is {`name`, `entry`} and that entry is present in the cache, so technically it isn't _unknown_, however, it is fair to assume that user wants to update `opaque payload` associated with {name, entry} pair, I deliberately didn't add it thinking `callback` would generate deterministic `payload` for a given {name, entry} pair. But it seems there could be valid usecases to update just the `payload`. I'm fine handling that usecase as well.

Reply:
If the cache has `[name: foo, ID: 1]` and `[name: bar, ID: 2]`, then it would reject the entry of `[name: bar, ID: 1]` even though there is no stored entry for `[name: bar, ID: 1]`. That's really what I meant by unknown, even though technically there is an outdated entry for ID 1 already present.

Reply:
That's interesting example; in my mind I always thought `name` needs to be associated with `id`; but maybe thats not the case, incase they are independent entity, then visualizing TenantInfo in-general gets more difficult, imho. 

I understand that user is responsible to assign those values, but, using fields interchangeably could lead to confusion in some scenarios, especially when `name` and/or `id` are interpreted to extract some metadata information. Not sure if there is rules or enforcements possible.

Reply:
This is where renames come into play. Imagine the following sequence:

1. Create tenant `foo` (assigned ID 1)
2. Create tenant `bar` (assigned ID 2)
3. The above two tenants are cached
4. Delete tenant `bar`
5. Rename tenant `foo` -> `bar`

Now we have a tenant `bar` with ID 1 and the two cached entries that overlap it.

Reply:
Thanks for sharing above example, it makes a lot of sense. 

I've incorporate the changes to the do the following:
1. Handle the scenario discussed above with appropriate entry cleanup. Also leveraged above scenario to add as code documentation (better code readability)
2. Update test actor to exercise above mentioned usecase.

Thanks